src/SourceGenerators/Uno.UI.SourceGenerators.netcore.Tests/MetadataUpdates/HotReloadWorkspaceProvider.cs

```diff
@@ -0,0 +1,345 @@
+	private readonly bool _isDebugCompilation;
+	private Dictionary<string, Dictionary<string, string>> _sourceFiles = new();
+	private Dictionary<string, Dictionary<string, string>> _additionalFiles = new();
+	private Solution? _currentSolution;
+	private WatchHotReloadService? _hotReloadService;
+
+	public HotReloadWorkspace(bool isDebugCompilation)
+	{
+		_isDebugCompilation = isDebugCompilation;
+		_baseWorkFolder = Path.Combine(Path.GetDirectoryName(typeof(HotReloadWorkspace).Assembly.Location)!, "work");
+		Directory.CreateDirectory(_baseWorkFolder);
+
+		Environment.SetEnvironmentVariable("Microsoft_CodeAnalysis_EditAndContinue_LogDir", _baseWorkFolder);
+	}
+
+	public void SetSourceFile(string project, string fileName, string content)
+	{
+		if (_sourceFiles.TryGetValue(project, out var files))
+		{
+			files[fileName] = content;
+		}
+		else
+		{
+			_sourceFiles[project] = new() {
+				[fileName] = content
+			};
+		}
+		var basePath = Path.Combine(_baseWorkFolder, project);
+		var filePath = Path.Combine(basePath, fileName);
+		Directory.CreateDirectory(basePath);
+		File.WriteAllText(filePath, content, Encoding.UTF8);
+
+		if(_currentSolution is not null)
+		{
+			var documents = _currentSolution
+				.Projects
+				.SelectMany(p => p.Documents)
+				.First(d => d.FilePath == filePath);
+
+			_currentSolution = _currentSolution.WithDocumentText(
+				documents.Id,
+				CSharpSyntaxTree.ParseText(content, encoding: Encoding.UTF8).GetText());
+		}
+	}
+
+	public void SetAdditionalFile(string project, string fileName, string content)
+	{
+		if (_additionalFiles.TryGetValue(project, out var files))
+		{
+			files[fileName] = content;
+		}
+		else
+		{
+			_additionalFiles[project] = new()
+			{
+				[fileName] = content
+			};
+		}
+		
+		var basePath = Path.Combine(_baseWorkFolder, project);
+		Directory.CreateDirectory(basePath);
+		var filePath = Path.Combine(basePath, fileName);
+		File.WriteAllText(filePath, content, Encoding.UTF8);
+
+		if (_currentSolution is not null)
+		{
+			var documents = _currentSolution
+				.Projects
+				.SelectMany(p => p.AdditionalDocuments)
+				.First(d => d.FilePath == filePath);
+
+			_currentSolution = _currentSolution.WithAdditionalDocumentText(
+				documents.Id,
+				CSharpSyntaxTree.ParseText(content, encoding: Encoding.UTF8).GetText());
+		}
+	}
+
+	public async Task Initialize(CancellationToken ct)
+	{
+		TaskCompletionSource<bool> taskCompletionSource = new();
+
+		var workspace = new AdhocWorkspace();
+
+		workspace.WorkspaceFailed += (_sender, diag) =>
+		{
+			if (diag.Diagnostic.Kind == WorkspaceDiagnosticKind.Warning)
+			{
+				Console.WriteLine($"MSBuildWorkspace warning: {diag.Diagnostic}");
+			}
+			else
+			{
+				taskCompletionSource.TrySetException(new InvalidOperationException($"Failed to create MSBuildWorkspace: {diag.Diagnostic}"));
+			}
+		};
+
+		var currentSolution = workspace.CurrentSolution;
+		var frameworkReferences = BuildFrameworkReferences();
```

Comment:
Can this be replaced with something like `await ReferenceAssemblies.Default.ResolveAsync()`?

Also if we don't care about the version of Uno references below we can then replace both lines with `await ReferenceAssemblies.Default.AddPackages(ImmutableArray.Create(new PackageIdentity("Uno.UI", "4.4.20"))).ResolveAsync()`

Reply:
We do care about the versions as new APIs are needed. I'm trying to find the right balance to get it working.